src/libraries/System.IO.Hashing/src/System/IO/Hashing/Crc32.cs

```diff
@@ -0,0 +1,168 @@
+        {
+            if (source is null)
+                throw new ArgumentNullException(nameof(source));
+
+            return Hash(new ReadOnlySpan<byte>(source));
+        }
+
+        /// <summary>
+        ///   Computes the CRC-32 hash of the provided data.
+        /// </summary>
+        /// <param name="source">The data to hash.</param>
+        /// <returns>The CRC-32 hash of the provided data.</returns>
+        public static byte[] Hash(ReadOnlySpan<byte> source)
+        {
+            byte[] ret = new byte[Size];
+            StaticHash(source, ret);
+            return ret;
+        }
+
+        /// <summary>
+        ///   Attempts to compute the CRC-32 hash of the provided data into the provided destination.
+        /// </summary>
+        /// <param name="source">The data to hash.</param>
+        /// <param name="destination">The buffer that receives the computed hash value.</param>
+        /// <param name="bytesWritten">
+        ///   On success, receives the number of bytes written to <paramref name="destination"/>.
+        /// </param>
+        /// <returns>
+        ///   <see langword="true"/> if <paramref name="destination"/> is long enough to receive
+        ///   the computed hash value (4 bytes); otherwise, <see langword="false"/>.
+        /// </returns>
+        public static bool TryHash(ReadOnlySpan<byte> source, Span<byte> destination, out int bytesWritten)
+        {
+            if (destination.Length < Size)
+            {
+                bytesWritten = 0;
+                return false;
+            }
+
+            bytesWritten = StaticHash(source, destination);
+            return true;
+        }
+
+        /// <summary>
+        ///   Computes the CRC-32 hash of the provided data into the provided destination.
+        /// </summary>
+        /// <param name="source">The data to hash.</param>
+        /// <param name="destination">The buffer that receives the computed hash value.</param>
+        /// <returns>
+        ///   The number of bytes written to <paramref name="destination"/>.
+        /// </returns>
+        public static int Hash(ReadOnlySpan<byte> source, Span<byte> destination)
+        {
+            if (destination.Length < Size)
+                throw new ArgumentException(SR.Argument_DestinationTooShort, nameof(destination));
+
+            return StaticHash(source, destination);
+        }
+
+        private static int StaticHash(ReadOnlySpan<byte> source, Span<byte> destination)
+        {
+            uint crc = InitialState;
+            crc = Update(crc, source);
+            BinaryPrimitives.WriteUInt32LittleEndian(destination, ~crc);
+            return Size;
+        }
+
+        private static uint Update(uint crc, ReadOnlySpan<byte> source)
+        {
+            for (int i = 0; i < source.Length; i++)
+            {
+                byte idx = (byte)crc;
+                idx ^= source[i];
+                crc = s_crcLookup[idx] ^ (crc >> 8);
+            }
```

Comment:
Would we be able to use any of the hardware intrinsics to speed this up if we had a .NET 6+ build?

Reply:
+1 to this. If we make this an inbox binary (which I think our partners would really want us to do), it'd make sense to intrinsicify these code paths.

Reply:
> If we make this an inbox binary (which I think our partners would really want us to do)

If I understand correctly, our thoughts for this library is "pure OOB".

Reply:
Why would the lib need to be inbox to benefit from hardware intrinsics? Doesn't it just need to compile against net6.0 which and could be included in the package?

Reply:
It only needs to compile against at least 3.0 to get acceleration on x86/x64 and 5.0 for acceleration on ARM64.

In particular, it would be nice if we just put the `CRC32C` algorithm under: https://github.com/dotnet/runtime/issues/2036 (which is approved just NYI) and had this API forward to it.
You can then also accelerate plain `CRC32` on Arm64, but that requires explicit intrinsics usage outside an xplat helper method.

Reply:
Well, but those are CRC32C, not CRC32.  If there's a sensible way to get from CRC32C to CRC32, then sure.

Reply:
For regular CRC32 (that is the variant where `0x04C11DB7` is the polynomial) its only available to accelerate on ARM64 (at least via a direct instruction, although technically also for Arm32, we just don't have intrinsic support here).

For CRC32-C (that is the variant where `0x1EDC6F41` is the polynomial) its available to x86/x64 and ARM64.